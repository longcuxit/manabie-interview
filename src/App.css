.App::before {
  position: fixed;
}
.App > .Loader-spinner {
  position: fixed;
}

.disabled {
  pointer-events: none;
  opacity: 0.5;
}

.reverse > .dropdown-menu[data-bs-popper] {
  top: auto;
  bottom: 100%;
  left: 0;
  margin-bottom: var(--bs-dropdown-spacer);
}

.input-group.collapsed {
  width: 40px;
}

.input-group.collapsed input {
  width: 0;
  padding-left: 0;
  padding-right: 0;
}

.container {
  max-width: 600px !important;
}

.form-group {
  position: relative;
}

.form-group > .form-control::placeholder {
  opacity: 0;
  transition: all ease-out 0.2s;
}
.form-group > .form-control:focus::placeholder {
  opacity: 1;
}

.form-group > .form-control + .form-label {
  position: absolute;
  left: 1rem;
  top: 50%;
  transform: translateY(-50%);
  padding: 0.25rem;
  transition: all ease-out 0.2s;
  border-radius: 4px;
  line-height: 1;
}

.form-group > .form-control:focus + .form-label,
.form-group > .form-control:not(:placeholder-shown) + .form-label {
  top: -3px;
  transform: translateY(-50%) scale(0.95);
  background-color: #fff;
}
