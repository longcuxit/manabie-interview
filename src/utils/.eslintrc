{
  "extends": "../../.eslintrc",
  "rules": {
    "no-restricted-imports": [
      "error",
      {
        "patterns": [
          "components/**",
          "models/**",
          "store/**",
          "pages",
          "pages/**",
          "service",
          "!./**"
        ]
      }
    ]
  }
}