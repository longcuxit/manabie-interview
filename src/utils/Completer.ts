enum CompleterStatus {
  running,
  resolved,
  rejected,
}

export interface Completer<T = void> extends Promise<T> {
  readonly completed: boolean;
  readonly rejected: boolean;
  readonly resolve: (value: T | PromiseLike<T>) => void;
  readonly reject: (error?: any) => void;
}

export class Completer<T = void> {
  constructor() {
    let resolve!: (value: T | PromiseLike<T>) => void;
    let reject!: (error?: any) => void;

    const promise = new Promise<T>((next, error) => {
      resolve = next;
      reject = error;
    }) as Completer<T>;

    let status = CompleterStatus.running;

    const checkBadStack = () => {
      if (status) {
        throw Error("Completer already completed!");
      }
    };

    Object.defineProperties(promise, {
      completed: { get: () => Boolean(status) },
      rejected: { get: () => status === CompleterStatus.rejected },
      resolve: {
        value(value: T | PromiseLike<T>) {
          checkBadStack();
          status = CompleterStatus.resolved;
          resolve(value);
        },
        writable: false,
      },
      reject: {
        value(error: any) {
          checkBadStack();
          status = CompleterStatus.rejected;
          reject(error);
        },
        writable: false,
      },
    });

    return promise;
  }
}
