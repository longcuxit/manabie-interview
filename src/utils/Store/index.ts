export * from "./type";
export * from "./Notifier";
export * from "./Subscriber";
export * from "./Store";
