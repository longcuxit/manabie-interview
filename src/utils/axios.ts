import axios from "axios";

const ins = axios.create({
  baseURL: "http://localhost:5050",
  timeout: 10000,
});

export default ins;
