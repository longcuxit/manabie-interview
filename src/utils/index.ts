export * from "./polyfill";
export * from "./Completer";

export function debounce<T, A extends any[] = never>(
  call: (this: T, ...args: A) => void,
  delay = 100
) {
  let timeOut: NodeJS.Timeout;
  let startTime = 0;
  return function (this: T, ...args: A) {
    const currentTime = Date.now();
    if (currentTime - startTime > delay) {
      startTime = currentTime;
      call.apply(this, args);
    } else {
      clearTimeout(timeOut);
      timeOut = setTimeout(() => {
        startTime = Date.now();
        call.apply(this, args);
      }, Math.max(startTime - currentTime + delay, 0));
    }
  };
}

export function createLocalStorage<T>(name: string, initial: T) {
  if (localStorage.getItem(name) === null)
    localStorage.setItem(name, JSON.stringify(initial));

  return {
    get() {
      return JSON.parse(localStorage.getItem(name)!);
    },
    set(value: T) {
      localStorage.setItem(name, JSON.stringify(value));
    },
  };
}
