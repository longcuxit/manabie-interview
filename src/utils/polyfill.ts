declare global {
  interface ObjectConstructor {
    keys<O>(object: O): (keyof O)[];
  }

  interface PromiseConstructor {
    delay(time: number): Promise<void>;
  }
}

Promise.delay = (time) => new Promise((next) => setTimeout(next, time));

export {};
