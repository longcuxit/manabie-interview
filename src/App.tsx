import { BrowserRouter } from "react-router-dom";

import "bootstrap/dist/css/bootstrap.min.css";

import { AsyncModal } from "components/AsyncModal";

import Pages from "./pages";

import "./App.css";
import { LoaderWrapper, usePushLoader } from "components/Loader";
import { useEffect, useState } from "react";
import Service from "service";

function App() {
  const [connected, setConnected] = useState(false);
  const pushLoader = usePushLoader();

  useEffect(() => {
    pushLoader(Service.connect()).then(setConnected);
  }, [pushLoader]);

  if (!connected) return null;

  return (
    <AsyncModal>
      <BrowserRouter basename={process.env.PUBLIC_URL}>
        <Pages />
      </BrowserRouter>
    </AsyncModal>
  );
}

export default function AppLoader() {
  return (
    <LoaderWrapper as="main" className="App">
      <App />
    </LoaderWrapper>
  );
}
