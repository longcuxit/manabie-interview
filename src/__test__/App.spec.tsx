import { render, act } from "@testing-library/react";
import { ReactNode } from "react";
import Service from "service";
import "utils";

import App from "../App";

jest.mock("../pages", () => () => <div className="pages">Pages</div>);
jest.mock("service", () => ({ connect: jest.fn() }));
jest.mock("components/Loader/Wrapper", () => ({
  LoaderWrapper: ({ children }: { children: ReactNode }) => <>{children}</>,
}));

const mockConnect = Service.connect as jest.Mock;

describe("App", () => {
  it("should connected service & match snapshot", async () => {
    mockConnect.mockResolvedValueOnce(true);
    const { container } = render(<App />);
    await act(() => Promise.resolve());

    expect(container.querySelector(".pages")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("should empty when can not connect service", async () => {
    mockConnect.mockResolvedValueOnce(false);
    const { container } = render(<App />);
    await act(() => Promise.resolve());

    expect(container.querySelector(".pages")).not.toBeInTheDocument();
  });
});
