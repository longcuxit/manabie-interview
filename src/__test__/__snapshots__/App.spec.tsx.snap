// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should connected service & match snapshot 1`] = `
<div>
  <div
    class="pages"
  >
    Pages
  </div>
</div>
`;
